src/MPQEditor/MPQCircleSelector.ts

```diff
@@ -0,0 +1,319 @@
+      args.document.uri.toString()
+    );
+    if (selPanel) {
+      selPanel.onForwardSelection(args.names);
+    }
+  }
+
+  private constructor(
+    panel: vscode.WebviewPanel,
+    extensionUri: vscode.Uri,
+    args: MPQSelectionCmdOpenArgs,
+    handler: MPQSelectionEvent | undefined
+  ) {
+    super(extensionUri, panel.webview);
+
+    // let parsedPath = path.parse(args.docPath);
+    // let fileBase = path.join(parsedPath.dir, parsedPath.name);
+
+    this._panel = panel;
+    this._ownerId = args.document.uri.toString();
+    this._document = args.document;
+    this._modelPath = args.modelPath;
+    this._mpqEventHandler = handler;
+    this._lastSelected = args.names;
+
+    // Listen for when the panel is disposed
+    // This happens when the user closes the panel or when the panel is closed programmatically
+    this._panel.onDidDispose(() => this.dispose(), null, this._disposables);
+
+    // Update the content based on view changes
+    this._panel.onDidChangeViewState(
+      () => {
+        if (this._panel.visible) {
+          // NOTE if we call this.update(), it'll reload the model which may take time.
+          // TODO call conditional this.update() when necessary.
+          // this.update();
+        }
+      },
+      null,
+      this._disposables
+    );
+
+    this.initGraphCtrl(this._modelPath, this);
+    this.setMode("selector");
+  }
+
+  public dispose() {
+    this.disposeGraphCtrl();
+
+    MPQSelectionPanel.panels.forEach((selPan, index) => {
+      if (
+        this._ownerId === selPan._ownerId &&
+        this._modelPath === selPan._modelPath
+      ) {
+        MPQSelectionPanel.panels.splice(index, 1);
+      }
+    });
+
+    // Clean up our resources
+    this._panel.dispose();
+
+    while (this._disposables.length) {
+      const x = this._disposables.pop();
+      if (x) {
+        x.dispose();
+      }
+    }
+  }
+
+  /**
+   * CircleGraphEvent interface implementations
+   */
+  public onViewMessage(message: any) {
+    switch (message.command) {
+      case MessageDefs.selection:
+        this._lastSelected = message.names;
+        // we need to update the document, but not save to file.
+        // pass to owner to handle this.
+        if (this._mpqEventHandler) {
+          this._mpqEventHandler.onSelection(message.names, this._document);
+        }
+        break;
+      case MessageDefs.finishload:
+        this.onForwardSelection(this._lastSelected);
+        break;
+    }
+  }
+
+  public onForwardSelection(selection: any) {
+    let selections: string[] = [];
+    let items = selection as Array<string>;
+    for (let i = 0; i < items.length; i++) {
+      if (items[i].length > 0) {
+        selections.push(items[i]);
+      }
+    }
+    this.setSelection(selections);
+  }
+
+  public setTitle(title: string) {
+    this._panel.title = title;
+  }
+
+  public onLoadVisqInfo(visqPath: string) {
+    const fileData = fs.readFileSync(visqPath, { encoding: "utf8", flag: "r" });
```

Comment:
AFAIK, it's rather safer to use `vscode.workspace.fs.readFile` instead of fs.readFileSync. Related issue: https://github.com/Samsung/ONE-vscode/issues/746

Reply:
Ok. Thank you. I didn't know that.